### PR TITLE
[EM-2855] Add support for Atome payment method

### PR DESCRIPTION
**Pull request by Narum11**
Add `Atome` as a new payment method.

Screen Shot 2023-01-04 at 5 20 22 PM  Screen Shot 2023-01-04 at 5 23 58 PM   Screen Shot 2023-01-04 at 5 24 23 PM

### CODE REVIEW

File: OmiseSDK/SourcePaymentMethod.swift [hunk 1]
```diff
@@ -37,6 +37,57 @@ public enum PaymentInformation: Codable, Equatable {
...
+        ///
+        /// - Parameters:
+        ///   - phoneNumber:  The customers phone number
+        public init(phoneNumber: String,
+                name: String? = nil,
+                email: String? = nil,
+                shippingStreet: String? = nil,
+                shippingCity: String? = nil,
+                shippingCountryCode: String? = nil,
+                shippingPostalCode: String? = nil) {
```
Comment (hpariwat) [inline OmiseSDK/SourcePaymentMethod.swift:77]: I believe the shipping postal code, country code, and street are also mandatory.

Comment (Narum11) [reply OmiseSDK/SourcePaymentMethod.swift]: Updated require fields as https://github.com/omise/omise-ios/pull/184/commits/bc0a068b875d28319dd0240abad5e07221c1b291 ka. 🙏

File: OmiseSDK/SourcePaymentMethod.swift [hunk 2]
```diff
@@ -37,6 +37,51 @@ public enum PaymentInformation: Codable, Equatable {
...
+        public var type: String = OMSSourceTypeValue.atome.rawValue
+
+        /// The customers phone number. Contains only digits and has 10 or 11 characters
+        public let phoneNumber: String
+        public let name: String?
+        public let email: String?
+        public let shippingStreet: String?
+        public let shippingCity: String?
+        public let shippingCountryCode: String?
+        public let shippingPostalCode: String?
```
Comment (kan98) [inline OmiseSDK/SourcePaymentMethod.swift:55]: street, country and postal code are not optional right?

Comment (kan98) [reply OmiseSDK/SourcePaymentMethod.swift:55]: should we remove the question mark at the end? it means variable is nullable right?

Comment (Narum11) [reply OmiseSDK/SourcePaymentMethod.swift]: Yes, T? represents nullable value type. Fix this one as https://github.com/omise/omise-ios/pull/184/commits/9b8d40e9d71a50bb65c3ca74a6af9eb2817d2ac8. Thank you ka. :)

File: OmiseSDK/AtomeFormViewController.swift [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            object: nil
+        )
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(keyboardWillHide(_:)),
+            name: NotificationKeyboardWillHideFrameNotification,
+            object: nil
+        )
+
+        phoneNumberTextField.validator = try? NSRegularExpression(pattern: "\\d{10,11}\\s?", options: [])
```
Comment (jutinapas) [inline OmiseSDK/AtomeFormViewController.swift:102]: Why do we have to have `\\s?` na? 🤔

File: OmiseSDK/AtomeFormViewController.swift [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        
+        guard let street = shippingStreetTextField.text?.trimmingCharacters(in: CharacterSet.whitespaces) else {
+            return
+        }
+        
+        guard let city = shippingCityTextField.text?.trimmingCharacters(in: CharacterSet.whitespaces) else {
+            return
+        }
+        
+        guard let country = shippingCountryCodeTextField.text?.trimmingCharacters(in: CharacterSet.whitespaces) else {
```
Comment (jutinapas) [inline OmiseSDK/AtomeFormViewController.swift:141]: I'm not sure if we have validation for this mandatory country code field, I remember that our Source API will throw an error if it’s not the 2-length characters 😂

File: OmiseSDK/AtomeFormViewController.swift (hunk 3 above)
Comment (Narum11) [reply OmiseSDK/AtomeFormViewController.swift:102]: In case user may have zero or one whitespace at the end, and we also trim phone string before assign value, somehow   we can allow only decimal digit. I will fix this one ka. thank you, bro! :joy:

File: OmiseSDK/AtomeFormViewController.swift (hunk 4 above)
Comment (Narum11) [reply OmiseSDK/AtomeFormViewController.swift:141]: Nice question ka, I will add this one too! 🙏

File: OmiseSDK/AtomeFormViewController.swift [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+            object: nil
+        )
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(keyboardWillHide(_:)),
+            name: NotificationKeyboardWillHideFrameNotification,
+            object: nil
+        )
+
+        phoneNumberTextField.validator = try? NSRegularExpression(pattern: "\\d{10,11}?", options: [])
```
Comment (hpariwat) [inline OmiseSDK/AtomeFormViewController.swift:107]: I'm wondering if it'd be confusing if we don't allow the "+" sign but we expect the country code in the field 🤔

Comment (Narum11) [reply OmiseSDK/AtomeFormViewController.swift:107]: I agree with you ka. :joy: 
I will add allow "+", and always add "+" if user fill in value without this one before send to api. 🙏

Comment (Narum11) [reply OmiseSDK/AtomeFormViewController.swift]: Please review again ka. I've changed a little bit from previous reply. This form only allow phone number which start with '+', and add phone number hint for guide user should enter '+' including country code.

File: OmiseSDK/AtomeFormViewController.swift [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+            object: nil
+        )
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(keyboardWillHide(_:)),
+            name: NotificationKeyboardWillHideFrameNotification,
+            object: nil
+        )
+
+        phoneNumberTextField.validator = try? NSRegularExpression(pattern: "^[0-9+]{0,1}+\\d{10,11}?", options: [])
```
Comment (jutinapas) [inline OmiseSDK/AtomeFormViewController.swift:115]: How about -> `(?:(?:\\+\\d{2})|0)?\\d{9}` (a plus sign following with 2 digits, or zero then 9 digits after). 
I'm not sure this will work 😂 but you can try this out krub

Comment (Narum11) [reply OmiseSDK/AtomeFormViewController.swift]: Changed as https://github.com/omise/omise-ios/pull/184/commits/833358db2d693cc567123cf9020a1b033544e3db ka. (also tested as you mention `a plus sign following with 2 digits, or zero then 9 digits after.`

Comment (jutinapas) [reply OmiseSDK/AtomeFormViewController.swift]: Cool! 👍

**Review by jutinapas — APPROVED** [review]

**Review by kan98 — APPROVED** [review]

**Review by hpariwat — APPROVED** [review]
